subprojects/file-collections/src/main/java/org/gradle/api/internal/file/AbstractFileCollection.java

```diff
@@ -105,17 +104,9 @@ protected void appendContents(TreeFormatter formatter) {
     @Override
     public final TaskDependency getBuildDependencies() {
         assertCanCarryBuildDependencies();
-        return new AbstractTaskDependency() {
-            @Override
-            public String toString() {
-                return "Dependencies of " + getDisplayName();
-            }
-
-            @Override
-            public void visitDependencies(TaskDependencyResolveContext context) {
-                context.add(AbstractFileCollection.this);
-            }
-        };
+        DefaultTaskDependency result = taskDependencyFactory.visitingDependencies(context -> context.add(AbstractFileCollection.this));
+        result.setToStringProvider(() -> "Dependencies of " + getDisplayName());
```

Comment:
🤔 I'm not so crazy about this one.

Reply:
Well, having meaningful `toString()` might still be useful in debugging. However, there was just this one `TaskDependency` subtype that had overridden `toString()`. Do you think keeping the ability to customize `toString()` in `DefaultTaskDependency` is not worth it?

Reply:
I agree having a meaningful `toString` is useful.

I guess I was expecting the lambda to be passed to the factory (or to a builder returned by the factory) instead of being set in the returned implementation but it's no big deal.